dlk/python/dlk/templates/src/matrix/shift_add.cpp

```diff
@@ -81,16 +80,70 @@ void matrix_shift_add(MatrixView<float, MatrixOrder::ColMajor>& buf,
 
               unsigned int j = 0;
 #ifdef USE_NEON
-              for (; j + 3 < oc; j += 4) {
-                float32x4_t b_ = vld1q_f32(b+j);
-                float32x4_t r_ = vld1q_f32(r+j);
-                float32x4_t r__ = vaddq_f32(b_, r_);
-                vst1q_f32(r+j, r__);
-              }
+              if (oc <= 8) {
+                for (; j + 3 < oc; j += 4) {
+                  float32x4_t b_ = vld1q_f32(b+j);
+                  float32x4_t r_ = vld1q_f32(r+j);
+                  float32x4_t r__ = vaddq_f32(b_, r_);
```

Comment:
wonder if we could come up with better variable naming than `r_` and `r__` ... :thinking:

Reply:
I think so too. Do you have some nice idea?

Reply:
In this context it is complicated, but for example
`r`, `b` -> `result_ptr`, `buf_ptr`
`r_`, `b_` -> `result_element`,  `buf_element`
`r__` -> ... it seems that we always store it immediately after assignment, so it may happen that do not create temporary varyable is actually better in this case...

so it becomes something like
```cpp
float32x4_t buffer_element = vld1q_f32(buffer_ptr+j);
float32x4_t result_element = vld1q_f32(result_ptr+j);
vstd1q_f32(result_ptr+j, vaddq_f32(buffer_element, result_element);
```

Reply:
Is the code correct? I am trying to implement software pipelining.

Reply:
discussed offline. :innocent:

Reply:
I re-thought about this problem.
`vst1q_f32(result_ptr+j, vaddq_f32(buffer_element, result_element));`
and
`r__ = vaddq_f32(buffer_element, result_element);
vst1q_f32(result_ptr+j, r__);`
are not equivalent in term of instruction scheduling.

But this is a trivial problem and performance deterioration is slight.

Reply:
The reason why they are not equivalent is not obvious...

Reply:
function(intrinsic) arguments have strong dependency. 
So, I think we have no room to re-schedule instructions.

So we might not be able to hide instruction latency. But a little.